proxy_map.go

```diff
@@ -0,0 +1,1023 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package hazelcast
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/hazelcast/hazelcast-go-client/internal"
+	"github.com/hazelcast/hazelcast-go-client/internal/cb"
+	"github.com/hazelcast/hazelcast-go-client/internal/event"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/codec"
+	iproxy "github.com/hazelcast/hazelcast-go-client/internal/proxy"
+	"github.com/hazelcast/hazelcast-go-client/predicate"
+	pubserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+type MapEntryListenerConfig struct {
+	Predicate          predicate.Predicate
+	IncludeValue       bool
+	Key                interface{}
+	NotifyEntryAdded   bool
+	NotifyEntryRemoved bool
+	NotifyEntryUpdated bool
+}
+
+type Map struct {
+	*proxy
+	refIDGenerator *iproxy.ReferenceIDGenerator
+	ctx            context.Context
```

Comment:
Context shouldn't be stored at the proxy level since the context should be provided at the operation level. Do we expect users to create separate proxies for each operation/chain of operations?

The same question applies to other proxies.

Reply:
`GetMapContext` doesn't create a proxy if it already exists. Current context implementation is experimental and used mainly for lock support. This is not the final design.

Reply:
The thing is that `GetMapContext` method is root level, so it looks a lot like `GetMap` and thus many users may use it as a way to obtain a proxy instance and cache it in their modules. The purpose of the method would be more obvious if it would be available on the `Map` struct itself (see `withContext` method).